File: js/portfolio.js
const Money = require('./money')

class Portfolio {
    constructor() {
        this.monies = new Array();
    }

    add(...monies) {
        this.monies = this.monies.concat(monies);
    }

    evaluate(bank, currency) {
        let failures = [];
        let total = this.monies.reduce((sum, money) => {
            try {
                let convertedMoney = bank.convert(money, currency);
                return sum + convertedMoney.amount;

            } catch (error) {
                failures.push(error.message);
                return sum;
            }
        }, 0);
        if (!failures.length) {
            return new Money(total, currency);
        }
        throw new Error("Missing exchange rate(s):[" + failures.join() + "]");
    }
}

module.exports = Portfolio

File: py/portfolio.py
from typing import Dict

from money import Money


class Portfolio:
    def __init__(self) -> None:
        self.monies = []
        self.exchange_rates: Dict[str, float] = {
            "EUR->USD": 1.2,
            "USD->KRW": 1100,
        }

    def add(self, *monies) -> None:
        self.monies.extend(monies)

    def evaluate(self, currency: float) -> Money:
        total = 0.0
        failures = []
        for money in self.monies:
            try:
                total += self.convert(money, currency)
            except KeyError as error:
                failures.append(error)

        if not failures:
            return Money(total, currency)

        failure_message = ",".join(f.args[0] for f in failures)
        raise Exception(f"Missing exchange rate(s):[{failure_message}]")

    def convert(self, money: Money, currency: str) -> float:
        rate = f"{money.currency}->{currency}"

        if money.currency == currency:
            return money.amount

        return money.amount * self.exchange_rates[rate]
